backend/src/composite/split.rs

```diff
@@ -0,0 +1,194 @@
+
+    let mut statements_by_namespace: BTreeMap<String, Vec<StatementIdentifier>> = BTreeMap::new();
+    for statement in pil.source_order.clone() {
+        let statement = match &statement {
+            StatementIdentifier::Definition(name)
+            | StatementIdentifier::PublicDeclaration(name) => {
+                let new_namespace = get_namespace(name);
+                current_namespace = new_namespace;
+                Some(statement)
+            }
+            StatementIdentifier::Identity(i) => {
+                let identity = &pil.identities[*i];
+                let namespaces = referenced_namespaces(identity);
+
+                match namespaces.len() {
+                    0 => panic!("Identity references no namespace: {identity}"),
+                    1 => {
+                        assert!(namespaces.iter().next().unwrap() == &current_namespace);
+                        Some(statement)
+                    }
+                    _ => match identity.kind {
+                        IdentityKind::Plookup | IdentityKind::Permutation => {
+                            assert_eq!(
+                                referenced_namespaces(&identity.left).len(),
+                                1,
+                                "LHS of identity references multiple namespaces: {identity}"
+                            );
+                            assert_eq!(
+                                referenced_namespaces(&identity.right).len(),
+                                1,
+                                "RHS of identity references multiple namespaces: {identity}"
+                            );
+                            log::debug!("Skipping connecting identity: {identity}");
+                            None
+                        }
+                        _ => {
+                            panic!("Identity references multiple namespaces: {identity}");
+                        }
+                    },
+                }
+            }
+        };
+
+        if let Some(statement) = statement {
+            statements_by_namespace
+                .entry(current_namespace.clone())
+                .or_default()
+                .push(statement);
+        }
+    }
+    statements_by_namespace
+}
+
+/// Given a PIL and a list of statements, returns a new PIL that only contains the
+/// given subset of statements.
+/// Returns None if there are no identities in the subset of statements.
+fn build_machine_pil<F: FieldElement>(
+    pil: Analyzed<F>,
+    statements: Vec<StatementIdentifier>,
+) -> Option<Analyzed<F>> {
+    // TODO: After #1488 is fixed, we can implement this like so:
+    // let pil = Analyzed {
+    //     source_order: statements,
+    //     ..pil.clone()
+    // };
+    // let parsed_string = powdr_parser::parse(None, &pil.to_string()).unwrap();
+    // let pil = powdr_pil_analyzer::analyze_ast(parsed_string);
+
+    // HACK: Replace unreferenced identities with 0 = 0, to avoid having to re-assign IDs.
```

Comment:
I guess we need an "ID reassigning" procedure anyway for the other backends?

Reply:
Ah I guess also not needed if re-parsing works?

Reply:
Yeah, exactly, then this code goes away.